File: remove-all-occurrences-of-a-substring/remove-all-occurrences-of-a-substring.java
class Solution {
    public String removeOccurrences(String s, String part) {
        
        if(part.length() == 0 || s.length() == 0){
            return s;
        }
        
        List<Character> list = new LinkedList<>();
        int p = part.length();
        
        for(char c : s.toCharArray()){
            
            if(list.size() == 0 && part.length() == 1 && c != part.charAt(0)){
                list.add(c);
            }
            else{
                if(list.size() + 1 >= p && c == part.charAt(p - 1)){
                    int n = list.size() - 1;
                    int j = p - 2;
                    while(n >= 0 && j >= 0 && list.get(n) == part.charAt(j)){
                        n--;
                        j--;
                        
                    }
                    if(j == -1){
                        j = list.size() - 1;
                        while(j > n){
                            list.remove(j--);
                        }
                    }
                    else{
                        list.add(c);
                    }
                }
                else{
                    list.add(c);
                }
            }
        }
        
        s = "";
        
        for(int i = 0; i < list.size(); i++){
            s += Character.toString(list.get(i));
        }
        return s;  
    }
}

File: remove-all-occurrences-of-a-substring/remove-all-occurrences-of-a-substring.cpp
class Solution {
public:
    string removeOccurrences(string s, string part) {
        
        vector<char> list;
        int n = part.length();
        for(char& c : s ){
            if(list.size() == 0 && part.length() == 1 && c != part[0]){
                list.push_back(c);
            }
            else{
                if(list.size() + 1 >= n && c == part[n - 1]){
                    int j = list.size() - 1;
                    int p = part.length() - 2;
                    while(j >= 0 && p >= 0 && part[p] == list[j]){
                        p--;
                        j--;
                    }
                    if(p == -1){
                        //cout<< list[j] << endl;
                        p = list.size() - 1;
                        
                        while(p > j){
                            list.pop_back();
                            p--;
                        }
                    }
                    else{
                        list.push_back(c);
                    }
                }
                else{
                    list.push_back(c);
                }
            }
        }
        
        s = "";
        for(int i = 0; i < list.size(); i++){
            s += list[i];
        }
        return s;
    }
};